File: linkedlist/design_linkedlist/design_linkedlist.go
package designlinkedlist

type MyLinkedList struct {
	dummy *Node
}

// 定义节点类型
type Node struct {
	Val  int
	Pre  *Node
	Next *Node
}

func Constructor() MyLinkedList {
	dummy := &Node{
		Val:  -1,
		Pre:  nil,
		Next: nil,
	}
	dummy.Next = dummy
	dummy.Pre = dummy
	return MyLinkedList{dummy}
}

func (this *MyLinkedList) Get(index int) int {
	head := this.dummy.Next
	for head != this.dummy && index > 0 {
		head = head.Next
		index--
	}

	// 索引无效
	if index > 0 {
		return -1
	}
	return head.Val
}

func (this *MyLinkedList) AddAtHead(val int) {
	dummy := this.dummy
	head := &Node{
		Val:  val,
		Pre:  dummy,
		Next: dummy.Next,
	}
	dummy.Next.Pre = head
	dummy.Next = head
}

func (this *MyLinkedList) AddAtTail(val int) {
	dummy := this.dummy
	rear := &Node{
		Val:  val,
		Pre:  dummy.Pre,
		Next: dummy,
	}

	dummy.Pre.Next = rear
	dummy.Pre = rear
}

// 在链表的第index个节点之前添加值
func (this *MyLinkedList) AddAtIndex(index int, val int) {
	head := this.dummy.Next
	for head != this.dummy && index > 0 {
		head = head.Next
		index--
	}
	if index > 0 {
		return
	}

	node := &Node{
		Val:  val,
		Pre:  head.Pre,
		Next: head,
	}
	head.Pre.Next = node
	head.Pre = node
}

func (this *MyLinkedList) DeleteAtIndex(index int) {
	if this.dummy == this.dummy.Next {
		return
	}

	head := this.dummy.Next
	/*
		为什么这里是head.Next而不是head
		比如目前是dummy -> 2，index=1，这种情况下就不行
	*/
	for head.Next != this.dummy && index > 0 {
		head = head.Next
		index--
	}
	if index == 0 {
		head.Pre.Next = head.Next
		head.Next.Pre = head.Pre
		head.Pre = nil
		head.Next = nil
	}
}

File: linkedlist/design_linkedlist/DesignLinkedlist.java
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2022-10-04 16:45:15
 * @LastEditTime: 2022-12-31 21:22:09
 */
package linkedlist.design_linkedlist;

/**
 * leetcode 707 middle. 设计链表
 * 设计链表的实现。您可以选择使用单链表或双链表。
 * 单链表中的节点应该具有两个属性：val 和 next。val 是当前节点的值，next 是指向下一个节点的指针/引用。
 * 如果要使用双向链表，则还需要一个属性 prev 以指示链表中的上一个节点。假设链表中的所有节点都是 0-index 的。
 */
public class DesignLinkedlist {

    ListNode dummy;
    public DesignLinkedlist() {
        dummy = new ListNode(-1);
        dummy.next = dummy;
        dummy.prev = dummy;
    }
    
    // index从0开始，是双向链表，到了最后还会转回来
    public int get(int index) {
        ListNode cur = dummy.next;
        while (index > 0 && cur != dummy) {
            cur = cur.next;
            index--;
        }

        return index == 0 ? cur.val : -1;
    }
    
    public void addAtHead(int val) {
        ListNode node = new ListNode(val);
        node.next = dummy.next;
        dummy.next.prev = node;
        node.prev = dummy;
        dummy.next = node;
    }
    
    public void addAtTail(int val) {
        ListNode node = new ListNode(val);
        node.prev = dummy.prev;
        dummy.prev.next = node;
        node.next = dummy;
        dummy.prev = node;
    }
    
    /*
     * 在链表中的第 index 个节点之前添加值为 val  的节点。
     * 如果 index 等于链表的长度，则该节点将附加到链表的末尾。
     * 如果 index 大于链表长度，则不会插入节点。
     * 如果 index 小于0，则在头部插入节点。
     * 
     */
    public void addAtIndex(int index, int val) {
        ListNode node = new ListNode(val);
        ListNode cur = dummy.next; 

        while (index > 0 && cur != dummy) {
            cur = cur.next;
            index--;
        }

        // 正常的话，出来就已经找到第index位置的节点了
        if (index > 0) {
            return;
        }

        node.prev = cur.prev;
        cur.prev.next = node;
        node.next = cur;
        cur.prev = node;
    }
    
    public void deleteAtIndex(int index) {
        ListNode cur = dummy.next; 

        while (index > 0 && cur != dummy) {
            cur = cur.next;
            index--;
        }

        // 正常的话，出来就已经找到第index位置的节点了
        if (index > 0) {
            return;
        }

        cur.prev.next = cur.next;
        cur.next.prev = cur.prev;
    }
}

class ListNode {
    int val; 
    ListNode next;
    ListNode prev; 

    public ListNode(int val) {
        this.val = val;
        this.next = null;
        this.prev = null;
    }
}